From ef82940238f02a5ba5f8f12303fd9032c6d8868d Mon Sep 17 00:00:00 2001
From: Matteo Facchinetti <matteo.facchinetti@sirius-es.it>
Date: Wed, 8 Feb 2017 17:44:09 +0100
Subject: [PATCH] bone-iio-helper: fix driver initialization order.

When you declared in the main dts the bone-iio-helper node and
driver is compiled statically into the kernel,
the helper driver is initialized before the ti_am335x_adc driver.
This generate a bug during boot initialization
because bone-iio-helper depends by ti_am335x_adc.

Fix it, changing the order of initialization of the bone-iio-helper
and placed among the last drivers to be init with late_initcall().

Signed-off-by: Matteo Facchinetti <matteo.facchinetti@sirius-es.it>
---
 drivers/misc/cape/beaglebone/bone-iio-helper.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/drivers/misc/cape/beaglebone/bone-iio-helper.c b/drivers/misc/cape/beaglebone/bone-iio-helper.c
index a919857..7384ea7 100644
--- a/drivers/misc/cape/beaglebone/bone-iio-helper.c
+++ b/drivers/misc/cape/beaglebone/bone-iio-helper.c
@@ -235,7 +235,18 @@ struct platform_driver bone_iio_helper_driver = {
 	},
 };
 
-module_platform_driver(bone_iio_helper_driver);
+static int __init bone_iio_helper_init(void)
+{
+	return platform_driver_register(&bone_iio_helper_driver);
+}
+
+static void __exit bone_iio_helper_exit(void)
+{
+	platform_driver_unregister(&bone_iio_helper_driver);
+}
+
+late_initcall(bone_iio_helper_init);
+module_exit(bone_iio_helper_exit);
 
 MODULE_AUTHOR("Matt Ranostay");
 MODULE_DESCRIPTION("Beaglebone IIO helper driver");
-- 
2.7.4

